board.go

```diff
@@ -0,0 +1,297 @@
+package rules
+
+import "math/rand"
+
+type BoardState struct {
+	Height  int32
+	Width   int32
+	Food    []Point
+	Snakes  []Snake
+	Hazards []Point
+}
+
+// NewBoardState returns an empty but fully initialized BoardState
+func NewBoardState(width, height int32) *BoardState {
+	return &BoardState{
+		Height:  height,
+		Width:   width,
+		Food:    []Point{},
+		Snakes:  []Snake{},
+		Hazards: []Point{},
+	}
+}
+
+// Clone returns a deep copy of prevState that can be safely modified inside Ruleset.CreateNextBoardState
+func (prevState *BoardState) Clone() *BoardState {
+	nextState := &BoardState{
+		Height:  prevState.Height,
+		Width:   prevState.Width,
+		Food:    append([]Point{}, prevState.Food...),
+		Snakes:  make([]Snake, len(prevState.Snakes)),
+		Hazards: append([]Point{}, prevState.Hazards...),
+	}
+	for i := 0; i < len(prevState.Snakes); i++ {
+		nextState.Snakes[i].ID = prevState.Snakes[i].ID
+		nextState.Snakes[i].Health = prevState.Snakes[i].Health
+		nextState.Snakes[i].Body = append([]Point{}, prevState.Snakes[i].Body...)
+		nextState.Snakes[i].EliminatedCause = prevState.Snakes[i].EliminatedCause
+		nextState.Snakes[i].EliminatedBy = prevState.Snakes[i].EliminatedBy
+	}
+	return nextState
+}
+
+// CreateDefaultBoardState is a convenience function for fully initializing a
+// "default" board state with snakes and food.
+// In a real game, the engine may generate the board without calling this
+// function, or customize the results based on game-specific settings.
+func CreateDefaultBoardState(width int32, height int32, snakeIDs []string) (*BoardState, error) {
+	initialBoardState := NewBoardState(width, height)
+
+	err := PlaceSnakesAutomatically(initialBoardState, snakeIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	err = PlaceFoodAutomatically(initialBoardState)
+	if err != nil {
+		return nil, err
+	}
+
+	return initialBoardState, nil
+}
+
```

Comment:
The only code that's really new past this point is `PlaceSnake` (a utility for adding a snake to the board). Everything else is lifted from `StandardRuleset` and renamed in a few cases to expose as public from the rules package.